diff --git a/node_modules/@semantic-release/npm/lib/verify-auth.js b/node_modules/@semantic-release/npm/lib/verify-auth.js
index d099337..2ed1182 100644
--- a/node_modules/@semantic-release/npm/lib/verify-auth.js
+++ b/node_modules/@semantic-release/npm/lib/verify-auth.js
@@ -81,6 +81,9 @@ async function verifyTokenAuth(registry, npmrc, context, pkgRoot) {
 }
 
 export default async function (npmrc, pkg, { pkgRoot }, context) {
+  if (context.options?.dryRun)
+    return;
+
   const registry = getRegistry(pkg, context);
 
   if (await oidcContextEstablished(registry, pkg, context)) {
